### PR TITLE
WIP: Initial (experimental) CoreOS support

### PR DESCRIPTION
**Pull request by justinsb**
* Detect CoreOS
* Move key manifests to code, to tolerate read-only mounts
* Misc refactorings so more code can be shared
* Change lots of ints to int32s in the models
* Run nodeup as a oneshot systemd service, rather than relying on
cloud-init behaviour which varies across distros

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1480)

### CODE REVIEW

**Comment by justinsb** [comment]
I can probably split this into smaller PRs, but comments appreciated!

**Comment by chrislovecnm** [comment]
Need to rebase ... I merged the other one first ... heh

**Comment by aledbf** [comment]
@justinsb @chrislovecnm exists a "procedure" to test new platforms? Or is just create cluster and verify everything works?
I can to help with this (and I really want CoreOS support in kops)

**Comment by justinsb** [comment]
Currently it's manual testing.  I would love to get a test matrix going, particularly so we have coverage of the different networking providers.  Even if it was randomized cc @zmerlynn .  I suspect this will require GCE support

**Comment by chrislovecnm** [comment]
Rebase please :)

File: cmd/nodeup/main.go [hunk 1]
```diff
@@ -64,17 +68,47 @@ func main() {
...
-			ModelDir:       models.NewAssetPath("nodeup"),
-		}
-		err := cmd.Run(os.Stdout)
-		if err == nil {
-			fmt.Printf("success")
-			os.Exit(0)
+		var err error
+		if install {
+			var command []string
+			for i := 0; i < len(os.Args); i++ {
```
Comment (chrislovecnm) [inline cmd/nodeup/main.go]: What is causing us to do this?  We have cobra, so I am confused.

File: cmd/nodeup/main.go [hunk 2]
```diff
@@ -64,17 +68,47 @@ func main() {
...
+				s := os.Args[i]
+				if s == "-install" || s == "--install" {
+					continue
+				}
+				if i == 0 {
+					// We could also try to evaluate based on cwd
+					s, err = os.Readlink("/proc/self/exe")
+					if err != nil {
+						glog.Fatalf("error reading /proc/self/exe link: %v", err)
```
Comment (chrislovecnm) [inline cmd/nodeup/main.go]: Could we have more context here with the error message?  The error message is that we cannot read `/proc/self/exe`, but why are we trying to read it.

File: cmd/nodeup/main.go [hunk 3]
```diff
@@ -64,17 +68,47 @@ func main() {
...
+					}
+				}
+				command = append(command, s)
+			}
+			i := bootstrap.Installation{
+				MaxTaskDuration: 5 * time.Minute,
```
Comment (chrislovecnm) [inline cmd/nodeup/main.go]: Should we have the Duration as a const?

File: nodeup/pkg/bootstrap/install.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+func (i *Installation) buildSystemdJob() *nodetasks.Service {
+	command := strings.Join(i.Command, " ")
+
+	serviceName := "kops-configuration.service"
+
+	manifest := &systemd.Manifest{}
+	manifest.Set("Unit", "Description", "Run kops bootstrap (nodeup)")
+	manifest.Set("Unit", "Documentation", "https://github.com/kubernetes/kops")
```
Comment (chrislovecnm) [inline nodeup/pkg/bootstrap/install.go]: Which element of kops are we trying to add documentation like for?

File: nodeup/pkg/model/codecs.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+	_ "k8s.io/kubernetes/pkg/api/install"
+)
+
+func encoder() runtime.Encoder {
+	// TODO: Which is better way to build yaml?
+	//yaml := json.NewYAMLSerializer(json.DefaultMetaFactory, k8sapi.Scheme, k8sapi.Scheme)
+
+	// TODO: Cache?
+	yaml, ok := runtime.SerializerInfoForMediaType(api.Codecs.SupportedMediaTypes(), "application/yaml")
```
Comment (chrislovecnm) [inline nodeup/pkg/model/codecs.go]: Do we have an issue for these TODOs?

File: nodeup/pkg/model/etcd.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		glog.Infof("Detected CoreOS; skipping etcd user installation")
+		return nil
+	}
+
+	// TODO: Do we actually use the user anywhere?
+
+	c.AddTask(&nodetasks.UserTask{
+		// TODO: Should we set a consistent UID in case we remount?
+		Name:  "user",
+		Shell: "/sbin/nologin",
```
Comment (chrislovecnm) [inline nodeup/pkg/model/etcd.go]: Does this always exist?  I am thinking yes, but just checking.

File: nodeup/pkg/model/kubeapiserver.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	addHostPathMapping(pod, container, "logfile", "/var/log/kube-apiserver.log", false)
+
+	pod.Spec.Containers = append(pod.Spec.Containers, *container)
+
+	return pod, nil
+}
+
+//// hasHostPath is used to check if the directory exists.  We use it because usually we don't want to set up a host-path
```
Comment (chrislovecnm) [inline nodeup/pkg/model/kubeapiserver.go]: Can we remove dead code?

File: nodeup/pkg/model/kubelet.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+		}
+		if asset == nil {
+			return fmt.Errorf("unable to locate asset %q", assetName)
+		}
+
+		t := &nodetasks.File{
+			Path:     b.kubeletPath(),
+			Contents: asset,
+			Type:     nodetasks.FileType_File,
+			Mode:     s("0755"),
```
Comment (chrislovecnm) [inline nodeup/pkg/model/kubelet.go]: const here?

File: nodeup/pkg/model/kubelet.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+	{
+		kubeconfig, err := b.buildKubeconfig()
+		if err != nil {
+			return err
+		}
+		t := &nodetasks.File{
+			Path:     "/var/lib/kubelet/kubeconfig",
+			Contents: fi.NewStringResource(kubeconfig),
+			Type:     nodetasks.FileType_File,
+			Mode:     s("0400"),
```
Comment (chrislovecnm) [inline nodeup/pkg/model/kubelet.go]: const here?

File: nodeup/pkg/model/kubelet.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+		c.AddTask(t)
+	}
+
+	c.AddTask(b.buildSystemdService())
+
+	return nil
+}
+
+func (b *KubeletBuilder) kubeletPath() string {
+	kubeletCommand := "/usr/local/bin/kubelet"
```
Comment (chrislovecnm) [inline nodeup/pkg/model/kubelet.go]: We have a bunch of strings here ... does it help us to make these constants? Also should we make the name of the component a const and then append the path?

File: pkg/apis/kops/cluster.go [hunk 11]
```diff
@@ -271,15 +271,13 @@ type KubeDNSConfig struct {
 	ServerIP string `json:"serverIP,omitempty"`
 }
 
-//
 //type MasterConfig struct {
```
Comment (chrislovecnm) [inline pkg/apis/kops/cluster.go]: Can we remove dead code?

File: upup/pkg/fi/nodeup/command.go [hunk 12]
```diff
@@ -181,27 +182,40 @@ func (c *NodeUpCommand) Run(out io.Writer) error {
...
 	glog.Infof("Config tags: %v", c.config.Tags)
 	glog.Infof("OS tags: %v", osTags)
 
-	modelContext := &model.NodeupModelContext{
-		Cluster:      c.cluster,
-		Distribution: distribution,
-		Architecture: model.ArchitectureAmd64,
+	tf, err := newTemplateFunctions(c.config, c.cluster, c.instanceGroup, nodeTags)
+	if err != nil {
+		return fmt.Errorf("error initializing: %v", err)
```
Comment (chrislovecnm) [inline upup/pkg/fi/nodeup/command.go]: > error initializing 

Intializing what?

File: upup/pkg/fi/nodeup/command.go [hunk 13]
```diff
@@ -181,27 +182,40 @@ func (c *NodeUpCommand) Run(out io.Writer) error {
...
+		IsMaster:      nodeTags.Has(TagMaster),
+		UsesCNI:       nodeTags.Has(tags.TagCNI),
+		Assets:        assets,
+		KeyStore:      tf.keyStore,
+		SecretStore:   tf.secretStore,
+	}
 
+	loader := NewLoader(c.config, c.cluster, assets, nodeTags)
 	loader.Builders = append(loader.Builders, &model.DockerBuilder{NodeupModelContext: modelContext})
+	loader.Builders = append(loader.Builders, &model.KubeletBuilder{NodeupModelContext: modelContext})
```
Comment (chrislovecnm) [inline upup/pkg/fi/nodeup/command.go]: Is there a way to not have to append?  Can we auto discover these methods?

File: nodeup/pkg/model/context.go [hunk 14]
```diff
@@ -19,10 +19,42 @@ package model
 import (
 	"k8s.io/kops/nodeup/pkg/distros"
 	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/upup/pkg/fi"
 )
 
 type NodeupModelContext struct {
```
Comment (hingstarne) [inline nodeup/pkg/model/context.go]: I assume this is a relict from cut and paste?

**Comment by justinsb** [comment]
Marking WIP while I continue to split this out into smaller PRs

File: cmd/nodeup/main.go (hunk 1 above)
Comment (justinsb) [inline cmd/nodeup/main.go]: This is creating a systemd process with the same args as we were called with.  I'll add a comment

File: cmd/nodeup/main.go (hunk 2 above)
Comment (justinsb) [inline cmd/nodeup/main.go]: Made more self-explanatory

File: cmd/nodeup/main.go (hunk 3 above)
Comment (justinsb) [inline cmd/nodeup/main.go]: No real reason to

File: nodeup/pkg/bootstrap/install.go (hunk 4 above)
Comment (justinsb) [inline nodeup/pkg/bootstrap/install.go]: Just cargo culting!

File: nodeup/pkg/model/codecs.go (hunk 5 above)
Comment (justinsb) [inline nodeup/pkg/model/codecs.go]: These are code-level TODOs; they don't warrant an issue.

File: nodeup/pkg/model/etcd.go (hunk 6 above)
Comment (justinsb) [inline nodeup/pkg/model/etcd.go]: I believe so

File: upup/pkg/fi/nodeup/command.go (hunk 12 above)
Comment (justinsb) [inline upup/pkg/fi/nodeup/command.go]: The hope is the error has the additional info.  We're initializing "stuff" here.  There's no point saying "error in newTemplateFunctions" IMO - it doesn't mean anything to the user.

File: upup/pkg/fi/nodeup/command.go (hunk 13 above)
Comment (justinsb) [inline upup/pkg/fi/nodeup/command.go]: I think it would introduce more bugs.  Explicit is better IMO.

File: nodeup/pkg/model/context.go (hunk 14 above)
Comment (justinsb) [inline nodeup/pkg/model/context.go]: I think I lost the line you were referring to @hingstarne , probably in a rebase?

File: pkg/model/resources/nodeup.go [hunk 15]
```diff
@@ -120,14 +120,9 @@ function download-release() {
...
-  # We run in the background to work around https://github.com/docker/docker/issues/23793
-  run-nodeup &
-}
-
-function run-nodeup() {
-  sleep 1
-  ( cd ${INSTALL_DIR}; ./nodeup --conf=/var/cache/kubernetes-install/kube_env.yaml --v=8 )
+  echo "Running nodeup"
+  # We can't run in the foreground because of https://github.com/docker/docker/issues/23793
+  ( cd ${INSTALL_DIR}; ./nodeup --install --conf=/var/cache/kubernetes-install/kube_env.yaml --v=8  )
```
Comment (aledbf) [inline pkg/model/resources/nodeup.go]: `--install` is not a valid flag

File: nodeup/pkg/bootstrap/install.go [hunk 16]
```diff
@@ -0,0 +1,111 @@
+package bootstrap
```
Comment (aledbf) [inline nodeup/pkg/bootstrap/install.go]: missing kubernetes header

**Comment by aledbf** [comment]
api-server manifest is missing. 
Besides that details updating nodeup the cluster is started without issues 
Using: 
```
kops create cluster \
  --name xxxxxx \
  --cloud aws \
  --master-zones $MASTER_ZONES \
  --node-count 3 \
  --zones $WORKER_ZONES \
  --master-size m3.medium \
  --node-size m4.xlarge \
  --ssh-public-key ~/.ssh/id_rsa.pub \
  --bastion \
  --networking calico \
  --topology private \
  --image 595879546273/CoreOS-stable-1235.9.0-hvm \
  --kubernetes-version "1.5.2"
```

**Comment by DocMerlin** [comment]
I am pretty excited by this.

**Comment by justinsb** [comment]
Indeed it can - covered by #1813 - thanks @aledbf :-)

**Pull request closed without merging**